### ISSUE TITLE
Brightness and Hue

### ISSUE THREAD

**Issue by Faisalthe01** [milo526/homebridge-tuya-web#11]
Hi, does you plugin fixed brightness and hue for bulbs ?

**Comment by thepotterfamily** [comment]
Not sure if this https://github.com/basdelfos/homebridge-tuya-web/pull/52 has anything to do with what you're describing?

**Comment by Faisalthe01** [comment]
This kind bloke was able to fix brightness and temperature for bulbs plus added cover accessory but a lot of other code has to go into config file to use it.

https://github.com/vitramir/homebridge-tuya-web/tree/feature/window-covering

**Comment by milo526** [comment]
Version 0.3.1 (#12 specifically) should have a fix in-place for the color/colour issue possibly fixing your problem if https://github.com/basdelfos/homebridge-tuya-web/pull/52 was indeed the problem you encountered.

If this issue persists feel free to re-open.

**Comment by milo526** [comment]
I just went out and got a lamp to test this with, there indeed seem to be some issues with brightness and hue.

```
[6/20/2020, 3:11:30 PM] [TuyaWebPlatform] [SET][Filament] Characteristic.Hue Error: Error: Invalid payload in response: {"payload":{},"header":{"payloadVersion":1,"code":"UnsupportedOperation"}}
    at /homebridge/node_modules/@milo526/homebridge-tuya-web/lib/tuyawebapi.js:169:28
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[6/20/2020, 3:14:23 PM] [TuyaWebPlatform] Error retrieving devices states TypeError: Cannot read property 'brightness' of undefined
    at LightAccessory.updateState (/homebridge/node_modules/@milo526/homebridge-tuya-web/lib/light_accessory.js:227:39)
    at LightAccessory.updateAccessory (/homebridge/node_modules/@milo526/homebridge-tuya-web/lib/base_accessory.js:103:14)
    at /homebridge/node_modules/@milo526/homebridge-tuya-web/index.js:96:52
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by Faisalthe01** [comment]
Yes, Vitramir fork has the fix for brightness and temperature not the hue though. I can also share the code he got me to add in my config file, if you want.

**Comment by dikodahan** [comment]
I am running an LED strip through the Tuya app (not Hue) and am getting this error:

```
[6/21/2020, 9:09:59 AM] [TuyaWebPlatform] Error retrieving devices states TypeError: Cannot read property 'brightness' of undefined
    at LightAccessory.updateState (/usr/lib/node_modules/homebridge-tuya-web/lib/light_accessory.js:222:43)
    at LightAccessory.updateAccessory (/usr/lib/node_modules/homebridge-tuya-web/lib/base_accessory.js:103:14)
    at /usr/lib/node_modules/homebridge-tuya-web/index.js:96:52
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

Seems like there may still be an issue with brightness. I am running the latest v.0.3.3 version of the plugin.
Thoughts!?

**Comment by milo526** [comment]
Hello could you please try and see if you can install the pre-release `0.3.4-0` by running `npm install homebridge-tuya-web@next` and see if that fixes the problem for you?

Please report back if this works for you or not 🙂

**Comment by dikodahan** [comment]
After updating to next version (v0.3.4) I am still getting the same:
```
[6/21/2020, 12:19:29 PM] [TuyaWebPlatform] Error retrieving devices states TypeError: Cannot read property 'brightness' of undefined
    at LightAccessory.updateState (/usr/lib/node_modules/@milo526/homebridge-tuya-web/lib/light_accessory.js:222:43)
    at LightAccessory.updateAccessory (/usr/lib/node_modules/@milo526/homebridge-tuya-web/lib/base_accessory.js:103:14)
    at /usr/lib/node_modules/@milo526/homebridge-tuya-web/index.js:96:52
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by milo526** [comment]
Could you then please run the discovery as explained [here](https://github.com/milo526/homebridge-tuya-web#how-to-check-whether-the-api-this-library-uses-can-control-your-device) and post the result that you get from that here (no need to open a new issue).

**Comment by dikodahan** [comment]
Just so we are on the same page.
I also have multiple outlets set up in Tuya app and working flawlessly with your plugin. I just added this LED strip today and it is the only one causing issues. The outlets still work without an issue.
Still need me to run the discovery?

**Comment by milo526** [comment]
Yes please, that will give me some information on the data structure that your LED strips sends and wants to receive. For example I know the brightness can be set on multiple locations and we use some vague techniques to determine which of the locations we need to get the brightness from. It seems like our techniques break with your device. Getting this info will help us improve that detection technique.

**Comment by dikodahan** [comment]
here you go (the data output part for the LED strip):
```
'data': {   'color_mode': 'colour', 'online': True, 'state': 'true'   },
    'dev_type': 'light',
    'ha_type': 'light',
    'icon': 'https://images.tuyaus.com/smart/program_category_icon/dj.png',
    'id': 'eb3aa814f1cbfebc48wqai',
    'name': 'Smart LED'},
```

**Comment by milo526** [comment]
Sadly your light strip does not support brightness through the home assistant API that this library uses (otherwise a `brightness` key would be present in your data object).
Do you see the LED strip in homekit? If so what controls do you have there?

**Comment by Faisalthe01** [comment]
@milo526 , this is the config i have in my config file for each and every bulb that Vitramir got me add it.

   ```
 {
                    "id": "",
                    "device_type": "light_temperature",
                    "config": {
                        "useCache": true,
                        "fromTuyaBrightness": [
                            {
                                "type": "parseInt"
                            },
                            {
                                "type": "divide",
                                "value": 10
                            },
                            {
                                "type": "round"
                            }
                        ],
                        "toTuyaBrightness": [
                            {
                                "type": "multiply",
                                "value": 0.9
                            },
                            {
                                "type": "add",
                                "value": 10
                            }
                        ],
                        "fromTuyaTemperature": [
                            {
                                "type": "parseInt"
                            },
                            {
                                "type": "subtract",
                                "value": 1000
                            },
                            {
                                "type": "divide",
                                "value": 35294
                            },
                            {
                                "type": "multiply",
                                "value": -1
                            },
                            {
                                "type": "add",
                                "value": 1
                            },
                            {
                                "type": "multiply",
                                "value": 360
                            },
                            {
                                "type": "add",
                                "value": 140
                            },
                            {
                                "type": "round"
                            }
                        ],
                        "toTuyaTemperature": [
                            {
                                "type": "subtract",
                                "value": 140
                            },
                            {
                                "type": "divide",
                                "value": 360
                            },
                            {
                                "type": "multiply",
                                "value": -1
                            },
                            {
                                "type": "add",
                                "value": 1
                            },
                            {
                                "type": "multiply",
                                "value": 9000
                            },
                            {
                                "type": "add",
                                "value": 1000
                            },
                            {
                                "type": "round"
                            }
                        ]
                    }
                },

```

**Comment by dikodahan** [comment]
@milo526 I'm OK with the fact that the specific device does not expose its brightness functionality. It is what it is.
But why do I need to get this error message from the plugin every 30 seconds?
If the device API does not expose it, why do you even display the error message?
Can the error message be displayed just once in the log and not every 30 seconds?

**Comment by milo526** [comment]
Yeah, that is clearly a bug. It should not show this error at all, it should just not allow you to update or get the brightness.

**Comment by camilogrifi** [comment]
I have the same bug, but with a bulb. On TUYA platform the light have the "brightness" property. I have problem with color too :(

**Comment by dikodahan** [comment]
> Yeah, that is clearly a bug. It should not show this error at all, it should just not allow you to update or get the brightness.

BTW, in HomeKit for me it does actually show the ability to change the brightness level, and it does work!

**Comment by Bad2dabone** [comment]
fratton/homebridge-tuya-web fork has fixed the Brightness problem and seems to work with led strips but the fork wont recognise my plugs,Where as your plugin does 
error with your plugin
[TuyaWebPlatform] Error retrieving devices states TypeError: Cannot read property 'brightness' of undefined
    at LightAccessory.updateState (C:\Users\uer\AppData\Roaming\npm\node_modules\@milo526\homebridge-tuya-web\lib\light_accessory.js:227:39)
    at LightAccessory.updateAccessory (C:\Users\uer\AppData\Roaming\npm\node_modules\@milo526\homebridge-tuya-web\lib\base_accessory.js:103:14)
    at C:\Users\uer\AppData\Roaming\npm\node_modules\@milo526\homebridge-tuya-web\index.js:96:52
    at processTicksAndRejections (internal/process/task_queues.js:97:5)

**Comment by milo526** [comment]
Since the devices I own don't seem to perform as described here, it is very hard for me to debug and thus fix this.

Could somebody please try to isolate the relevant changes made by fratton (and possibly cherry-pick them and make a PR)? That would be highly appreciated.

For everybody that is seeing the same problem, could you please run the discovery as explained [here](https://github.com/milo526/homebridge-tuya-web#how-to-check-whether-the-api-this-library-uses-can-control-your-device) (the "How to check whether the api this library uses can control your device" heading) and post the result that you get from that here (no need to open a new issue).

Please also state the following information
1) What information do you see in HomeKit ? on/off - brightness control - colour control - temperature control (shades of yellow/white)
2) What can you change in your tuya/smartlife app? on/off - brightness control - colour control - temperature control (shades of yellow/white) - anything else?
3) What functionality are you able to use in HomeKit? If you change one of the values, does that change anything on the device? Please note any action and the consequent changes, even if for example the brightness does change but is not as expected (i.e. too dark or too light) or the color does change but not to/from the expected color.

If at all possible please also post a listing to the device that you bought.

This will help me as well as anybody else who might pick this issue up, the more information we have, the faster we can resolve this issue.

**Comment by Bad2dabone** [comment]
hope this helps my discovery feedback
 Allso have found plugs are not updating (getswitchstate)in homekit if switched with smartlife app
Led strip in smartlife app can change brightness on/off Saturation and all colours 
Same is available in homekit just either not working or not in right order blue is red Brightness is 0% etc
Got credentials

Got devices
{   'header': {'code': 'SUCCESS', 'payloadVersion': 1},
    'payload': {   'devices': [   {   'data': {   'brightness': '255',
                                                  'color_mode': 'colour',
                                                  'online': True,
                                                  'state': 'false'},
                                      'dev_type': 'light',
                                      'ha_type': 'light',
                                      'icon': 'https://images.tuyaeu.com/smart/icon/1491806318_0.png',
                                      'id': '026660042cf4326afd5e',
                                      'name': 'Back Light'},
                                  {   'data': {'online': True, 'state': False},
                                      'dev_type': 'switch',
                                      'ha_type': 'switch',
                                      'icon': 'https://images.tuyaeu.com/smart/icon/ay1557739264731H9AZr/1582533575ab0ca246e14.png',
                                      'id': '22358400a4cf12d280e2',
                                      'name': 'Jug'},
                                  {   'data': {'online': True, 'state': False},
                                      'dev_type': 'switch',
                                      'ha_type': 'switch',
                                      'icon': 'https://images.tuyaeu.com/smart/icon/ay1557739264731H9AZr/1582533575ab0ca246e14.png',
                                      'id': '22358400500291598b98',
                                      'name': 'Plugcam1'}],
                   'scenes': []}}
!!! NOW REMOVE THIS FILE, SO YOUR CREDENTIALS (username, password) WON'T LEAK !!!


**Comment by Bad2dabone** [comment]
have also found another folk where everything seems to be working correctly but getting error code in homebridge logs
https://www.npmjs.com/package/homebridge-tuya-api
install with npm install -g homebridge-tuya-api
error log
[TuyaWebPlatform] Error retrieving devices states Error: No valid response from API
    at C:\Users\uer\AppData\Roaming\npm\node_modules\homebridge-tuya-api\lib\tuyawebapi.js:78:18
    at C:\Users\uer\AppData\Roaming\npm\node_modules\homebridge-tuya-api\lib\tuyawebapi.js:309:11
    at Request._callback (C:\Users\uer\AppData\Roaming\npm\node_modules\homebridge-tuya-api\lib\tuyawebapi.js:295:11)
    at Request.self.callback (C:\Users\uer\AppData\Roaming\npm\node_modules\homebridge-tuya-api\node_modules\request\request.js:185:22)
    at Request.emit (events.js:310:20)
    at Request.<anonymous> (C:\Users\uer\AppData\Roaming\npm\node_modules\homebridge-tuya-api\node_modules\request\request.js:1154:10)
    at Request.emit (events.js:310:20)
    at IncomingMessage.<anonymous> (C:\Users\uer\AppData\Roaming\npm\node_modules\homebridge-tuya-api\node_modules\request\request.js:1076:12)
    at Object.onceWrapper (events.js:416:28)
    at IncomingMessage.emit (events.js:322:22)
    at endReadableNT (_stream_readable.js:1187:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)

**Comment by charlesomer** [comment]
Would it not be possible to at least catch the exception so it doesn't cause a crash and disrupt the other devices. Maybe just provide a warning message instead if this is caught?

**Comment by milo526** [comment]
> Would it not be possible to at least catch the exception so it doesn't cause a crash and disrupt the other devices. Maybe just provide a warning message instead if this is caught?

Feel free to open a PR 😄

**Comment by Hans0lo** [comment]
i have the same issue, here's the output of the discovery:

```
{   'access_token': 'xxxxxxxxxxxxxxxxxxxxxxxxxxx',
    'expires_in': 864000,
    'refresh_token': 'xxxxxxxxxxxxxxxxxxxxxxxxxxx',
    'token_type': 'bearer'}
Getting devices
Got devices
{   'header': {'code': 'SUCCESS', 'payloadVersion': 1},
    'payload': {   'devices': [   {   'data': {   'brightness': '128',
                                                  'color_mode': 'colour',
                                                  'color_temp': 10000,
                                                  'online': True,
                                                  'state': 'true'},
                                      'dev_type': 'light',
                                      'ha_type': 'light',
                                      'icon': 'https://images.tuyaeu.com/smart/icon/ay1541056239985fDGjj/15620318474c719f1c5cf.png',
                                      'id': 'xxxxxxxxxxxxxxxxxxxxxxxxxxx',
                                      'name': 'LED Smart'}],
                   'scenes': []}}
```

**Comment by bwp91** [comment]
> i have the same issue, here's the output of the discovery:

Perhaps i could have a look into this.
Would you post your error message?
Unfortunately I've gone through the same battle against brightness, colours, hue etc with my ewelink plugin!

**Comment by Hans0lo** [comment]
```
[7/15/2020, 14:47:57] [TuyaWebPlatform] Error retrieving devices states TypeError: Cannot read property 'brightness' of undefined
    at LightAccessory.updateState (/usr/local/lib/node_modules/@milo526/homebridge-tuya-web/lib/light_accessory.js:227:39)
    at LightAccessory.updateAccessory (/usr/local/lib/node_modules/@milo526/homebridge-tuya-web/lib/base_accessory.js:103:14)
    at /usr/local/lib/node_modules/@milo526/homebridge-tuya-web/index.js:96:52
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by bwp91** [comment]
@Hans0lo are you still on v0.3.2 by any chance? And if so have you tried updating to v0.3.3?

**Comment by Hans0lo** [comment]
i am on 0.33 ATM
i now can for sure confirm that the error shown not up in white mode at all - only when in color mode

**Comment by bwp91** [comment]
Assuming then that the "brightness" property is only sent to Homebridge when the bulb is in white mode. 
When the bulb is in colour mode the brightness property isn't sent which is why the error is being thrown as the code doesn't know what it needs something it doesn't have.

`sudo npm install -g homebridge-tuya-web@next` from terminal within Homebridge should do it.

It looks like @milo526 has rewritten the light_accessory.js file since v0.3.3 so it makes little sense for me to suggest potential changes.

**Comment by milo526** [comment]
Thanks for looking into this @bwp91!

My own colour light in colour modes gives:

```
{
   "data":{
      "brightness":"119",
      "color_mode":"colour",
      "color_temp":5588,
      "online":true,
      "state":"true"
   },
   "dev_type":"light",
   "ha_type":"light",
   "icon":"https://images.tuyaeu.com/smart/icon/ay1541056239985fDGjj/15620318474c719f1c5cf.png",
   "id":"bfb7e99e6232be367bm54y",
   "name":"Kleur"
}
```
The same device in white mode yields

```
{
   "data":{
      "brightness":"166",
      "color_mode":"white",
      "color_temp":5588,
      "online":true,
      "state":"true"
   },
   "dev_type":"light",
   "ha_type":"light",
   "icon":"https://images.tuyaeu.com/smart/icon/ay1541056239985fDGjj/15620318474c719f1c5cf.png",
   "id":"bfb7e99e6232be367bm54y",
   "name":"Kleur"
}
```

So I believe your premise isn't exactly true. This also immediately shows one of the problems that this plugin has during development, there are 1000's of Tuya-based light bulbs and almost as many different payload types.

**Comment by bwp91** [comment]
> This also immediately shows one of the problems that this plugin has during development, there are 1000's of Tuya-based light bulbs and almost as many different payload types.

Completely get this! My package for ewelink (focusing on sonoff brand) is just a total case of if...thens.... even from the same brand some devices use "brightness" property and other devices use "bright". some between 0-100... some between 10-100. its a nightmare.

**Comment by bwp91** [comment]
But i think it seems silly to look at your code for 0.3.3 if you did a major rewrite of the code anyway?

**Comment by Hans0lo** [comment]
@bwp91  I am not sure if that worked correctly. Installation worked without error, but plugin page still shows 0.3.3. (after a reboot)

I then got some error shown and the pi went offline.

after a cold reboot I am back in and getting variants of the same old error:
```
[7/15/2020, 16:00:37] [TuyaWebPlatform] [GET][LED Smart] Characteristic.Brightness Error: TypeError: Cannot read property 'brightness' of undefined
    at /usr/local/lib/node_modules/@milo526/homebridge-tuya-web/lib/light_accessory.js:76:51
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[7/15/2020, 16:00:45] [Neo Lamp] ... saturation is currently 33
[7/15/2020, 16:00:45] [Neo Lamp] ... hue is currently 47
[7/15/2020, 16:00:45] [Neo Lamp] power is currently ON
[7/15/2020, 16:00:45] [Neo Lamp] brightness is currently at 75 %
[7/15/2020, 16:00:48] [Neo Lamp] brightness is currently at 75 %
[7/15/2020, 16:00:48] [Neo Lamp] ... saturation is currently 33
[7/15/2020, 16:00:48] [Neo Lamp] power is currently ON
[7/15/2020, 16:00:49] [Neo Lamp] ... hue is currently 47
[7/15/2020, 16:01:09] [TuyaWebPlatform] Error retrieving devices states TypeError: Cannot read property 'brightness' of undefined
    at LightAccessory.updateState (/usr/local/lib/node_modules/@milo526/homebridge-tuya-web/lib/light_accessory.js:227:39)
    at LightAccessory.updateAccessory (/usr/local/lib/node_modules/@milo526/homebridge-tuya-web/lib/base_accessory.js:103:14)
    at /usr/local/lib/node_modules/@milo526/homebridge-tuya-web/index.js:96:52
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by milo526** [comment]
I don't know what exact rewrite you are talking about; I'm rewriting the library to use TypeScript currently, the 3.4.0 code is already on master. The TypeScript rewrite shouldn't impede on this too much.

**Comment by Hans0lo** [comment]
now I can use the EVE slider to change brighness in both modes.

**Comment by bwp91** [comment]
> I don't know what exact rewrite you are talking about; I'm rewriting the library to use TypeScript currently, the 3.4.0 code is already on master. The TypeScript rewrite shouldn't impede on this too much.

Ah let me rephrase!
Assuming most people are still on 0.3.3 since .4 is a beta, there's not much point for me trying to solve the issue in the 0.3.3 code since the light-accessory file has been largely changed

**Comment by milo526** [comment]
> Ah let me rephrase!
> Assuming most people are still on 0.3.3 since .4 is a beta, there's not much point for me trying to solve the issue in the 0.3.3 code since the light-accessory file has been largely changed

Ah, check! I don't feel comfortable releasing that release now since I'll be gone the next week with little access to my laptop. Please remind me in 10 days if I haven't released by then 😉

**Comment by bwp91** [comment]
> @bwp91 I am not sure if that worked correctly.

@Hans0lo in homebridge plugins page which plugins do you have listed? Any chance you have two entries for "Homebridge Tuya Web"?

**Comment by Hans0lo** [comment]
@bwp91  arrgghh... yes, you are right, there were two entries, I just did not notice it, the newer one was sorted quite to the bottom.
I uninstalled 0.3.3 now.

if i should do any other tests or so, just let me know

**Comment by bwp91** [comment]
My input here was to try and sort out any errors that were crashing Homebridge (as 1) I know @milo526 has said somewhere that his time is taken up at the moment and 2) I personally know how annoying it can be when Homebridge has to restart indefinitely).
So very happy to help trying to sort errors between this package and Homebridge.
Are you at a point now where you still receiving errors?

With regards to functionality (eg enabling colour) -- that would be almost impossible for me to implement as I don't own any tuya devices anymore.

**Comment by Hans0lo** [comment]
since I only got these errors in color mode - and color mode is not working in 0.3.4 - i get no more errors 
also my homebridge only froze once (directly after updating), i did not get a crash loop .

the error was only cosmetic from my perspective, but i hoped to get the color/white/hue/saturation features working.
but thnx anyway  :)

**Comment by jamesmccarthy** [comment]
Reporting the same issue on Tuya lights with Colour and White (and scene) settings. 

**Behaviour in White mode**
- On/Off control in HomeKit works (if light is in white mode already)
- Light behaves as expected - white/warm light
- Brightness can be adjusted within HomeKit, but the scale is incorrect (10% on HomeKit = 0% on Tuya, 100% on HomeKit = 100% on Tuya, <10% in HomeKit = off in Tuya)
- Adjusting temperature in HomeKit switches light to Colour mode (see below)
- Adjusting brightness in Tuya App shows brightness % reading in HomeKit of 0% (which isn't correct)

**Behaviour in Colour mode**
- On/Off control in HomeKit works
- Light doesn't behave as expected - very dim blue light
- No change to temp/mode/hue/brightness settings in HomeKit changes the light
- Must reset the light to White mode in Tuya app to get it working again

**Behaviour in Scene mode**
- Same as Colour mode

_Accessory details attached for each mode_

[accessory_homekit.txt](https://github.com/milo526/homebridge-tuya-web/files/4974867/accessory_homekit.txt)

**Comment by milo526** [comment]
Well I didn't make the 10 days I said, it was 11.

Both color and color temperature seems to work perfectly for me with the new beta release, could you all please try installing `@milo526/homebridge-tuya-web@beta`. This is the new beta version of the package and it should fix these problems.
Furthermore it should reduce the amount of times you get the rate limiting errors.

I would advice you to disable the polling options by removing the polling key from your configuration file (or clearing the value in homebridge-config-ui-x)

**Comment by nokemono42** [comment]
@milo526 Hue and brightness work correctly now, but Homekit is very slow in updating the status. Also I was unable to add two lights to a group. (Which I could do in while on v0.3.3.)

**Comment by ShahrumAmiri** [comment]
Hi there -- I just wanted to weigh in that I just installed the 0.3.4 beta and it solved the same issues I was running into with an outdoor pluggable dimmer. It also fixed the fact that I was getting hue settings when the only control I should have had is brightness. Thank you!!

BTW: When 0.3.4 is released out of beta, will it auto-update since I now have the beta version installed? Or will I need to remove the beta and install the public release? Thanks!

**Comment by milo526** [comment]
This issue has been fixed in the [v0.4.2 release](https://github.com/milo526/homebridge-tuya-web/releases/tag/v0.4.2) if you are still experiencing this issue, or are experiencing other issues, please open a new bug report.

Thank you once again for this bug report, reports like these help make the plugin better!

@ShahrumAmiri The plugin should update automatically.

There are still some issues with regards to "requesting too quickly" if this is still an issue for you, please open a new issue or comment to another relevant open issue.
